lib/rubocop/cop/naming/memoized_instance_variable_name.rb

```diff
@@ -0,0 +1,62 @@
+# frozen_string_literal: true
+
+module RuboCop
+  module Cop
+    module Naming
+      # This cop checks for memoized methods whose instance variable name
+      # does not match the method name.
+      #
+      # @example
+      #   # bad
+      #
+      #   # Method foo is memoized using an instance variable that is
+      #   # not `@foo`. This can cause confusion and bugs.
+      #   def foo
+      #     @something ||= calculate_expensive_thing
+      #   end
+      #
+      #   # good
+      #
+      #   def foo
+      #     @foo ||= calculate_expensive_thing
+      #   end
+      #
+      #   # good
+      #
+      #   def foo
+      #     @foo ||= begin
+      #       calculate_expensive_thing
+      #     end
+      #   end
+      #
+      class MemoizedInstanceVariableName < Cop
+        MSG = 'Memoized variable does not match method name.'.freeze
+
+        private_class_method def self.node_pattern
+          memo_assign = '(or_asgn $(ivasgn _) _)'
+          memoized_at_end_of_method = "(begin ... #{memo_assign})"
+          "(def $_ _ {#{memo_assign} #{memoized_at_end_of_method}})"
+        end
+        def_node_matcher :memoized?, node_pattern
+
+        def on_def(node)
+          (method_name, ivar_assign) = memoized?(node)
+          return if matches?(method_name, ivar_assign)
+          add_offense(node, location: ivar_assign.source_range)
+        end
+
+        private def last_in_method(node)
+          node.body.begin_type? ? node.body.children.last : node.body
+        end
+
+        private def matches?(method_name, ivar_assign)
```

Comment:
Although we don't have one, I believe there is an open issue/discussion about adding a cop that sets a convention for access modifier assignment. There is an argument that inline modifier's are better because they are more explicit. But nevertheless, it's not a convention we use in the repo, so this would be an outlier.  Mind moving that `private` above your private instance methods instead of inline?